rts/idris_rts.c

```diff
@@ -556,62 +673,83 @@ VAL idris_strTail(VM* vm, VAL str) {
             root = root->info.str_offset->str;
         }
 
-        cl->info.str_offset->str = root;
-        cl->info.str_offset->offset = offset+idris_utf8_charlen(GETSTR(str));
+        cl -> info.str_offset -> str = root;
+        cl -> info.str_offset -> offset = offset + idris_utf8_charlen(string);
 
         return cl;
     } else {
-        char* nstr = GETSTR(str);
-        return MKSTR(vm, nstr+idris_utf8_charlen(nstr));
+        SizedString nstr = GETSTR(str);
+        unsigned int tailIdx = idris_utf8_charlen(nstr);
+        assert(tailIdx <= nstr.size);
+        Closure *cl = idris_allocate_string(nstr.size - tailIdx, 0);
+        // Use memcpy here instead of MKSTR because MKSTR assumes
+        // C-style strings, and 0 is a sensible Unicode character that
```

Comment:
`MKSTR` is really responsible for turning actual C strings into Idris RTS strings. It's used various places in the codegen, along with `MKINT` and friends. It might be feasible to plumb lengths through that, however, and that will probably be necessary in the end to allow `'\0'` as part of a string.